### ISSUE TITLE
Question - Weird prompt results

### ISSUE THREAD

**Issue by stellanhaglund** [karpathy/llama2.c#204]
I have trained a model with `max_seq_len=512` and examples in this format `some-text -> some-answer###` 
when I run the model without prompt i get new random data in the correct format but when I start with the prompt 
`"some-text ->" `I get only weird results and usually another -> with a good answer after it.

I'm not sure this is because of the prompting but it's strange that the outputs without a prompt works a lot better than with. 

I thought it could be because the examples where of different lengths so I tried padding them, but the tokenizer has -1 for pad_id so tried padding with the eos_id instead, but there doesn't seem to be any difference.

Does anyone have any insights here?

**Comment by stellanhaglund** [comment]
The model has a very good validation loss after 100k iterations, 0.35 so I guess it should be working good..

**Comment by sri9s** [comment]
Hi @stellanhaglund, do you actually have the < symbol in your dataset in the format you specified? What is your dataset size?

**Comment by stellanhaglund** [comment]
Yes I have it in the dataset.
It’s 900k rows with 250-500 tokens per row but padded to 512 tokens per row.

**Comment by stellanhaglund** [comment]
I tried most of the saved models along the training and its the same problem on all of them, so I don’t think its overfitted

**Comment by stellanhaglund** [comment]
from what I can tell the token encoder in run.c gives a result quite different from the python version I will try to hardcode the encoded prompt from the python version and see if I get better results

**Comment by stellanhaglund** [comment]
Ok so that was the issue, hardcoding the encoded prompt from the python version gives me perfect output, @karpathy do you have any ideas where the encoding differs from the original sentencepiece encoding?

**Comment by karpathy** [comment]
:| oh no... not really sadly, if someone has some cycles to track this down would welcome a PR fix

**Comment by stellanhaglund** [comment]
I tried adding sentencepiece in rust to llama2.rs and the prompting works good there, though a bit slower. 
Looked a bit at the c++ implementation and it doesn't look overly complicated, though it uses a hash-map, could probably be worked around with some looping though, but maybe not as performant. 

https://github.com/google/sentencepiece/blob/635fe8423a249b6e081aacd290d8aef7476c6a28/src/bpe_model.cc

**Comment by twobob** [comment]
the C++ version includes additional features like BPE-dropout and resegmentation over the current. I didn't find the python version on first glance? but perhaps it's some detail like that.

**Comment by stellanhaglund** [comment]
> the C++ version includes resegmentation which is not implemented in the current run.c. I didn't find the python version on first glance? but perhaps it's some detail like that.

The python lib is just bindnings to the c++ code

**Comment by twobob** [comment]
then it's likely the resegmentation.? Total guess on eyeballing the code. 
Other difference was dropout, which we dont use anyway.

**Comment by atamurad** [comment]
@stellanhaglund  can you test https://github.com/karpathy/llama2.c/pull/226 ?

I've added utf8 support and fixed one dummy whitespace (minor) issue.

Do you have other samples/cases where run.c and sentencepiece encoding output differs?

**Comment by atamurad** [comment]
> then it's likely the resegmentation.? Total guess on eyeballing the code.

@twobob I confirmed that the resegmentation part is required only if there are any type=unused tokens in the vocabulary, and there are none in tokenizer.model file.

**Comment by twobob** [comment]
> @stellanhaglund can you test #226 ?
> 
> I've added utf8 support and fixed one dummy whitespace (minor) issue.
> 
> Do you have other samples/cases where run.c and sentencepiece encoding output differs?

 spotted you added the utf8 support and did indeed wonder if that might fix it for him. Good job

**Comment by stellanhaglund** [comment]
Sorry for the late reply, I just tested this and it works! Awesome! ❤️

**Comment by stellanhaglund** [comment]
The prompt encoding seems quite a bit slower though? @atamurad

**Comment by atamurad** [comment]
> The prompt encoding seems quite a bit slower though? @atamurad 

How long is the prompt? I don't think utf8 is the reason.

If tokens/sec is lower it may be because of the top-p sampling added recently (it is in PR enabled by default, but now disabled in the master branch, but not in utf8 branch). Try with -p 0 to disable it.

**Comment by stellanhaglund** [comment]
🙏 I will try that out!
The prompt is about 180 tokens and takes about 1 sec or more to encode when I tried, but it’s almost instant with the master version.
Will try in a moment!

**Comment by stellanhaglund** [comment]
> > The prompt encoding seems quite a bit slower though? @atamurad
> 
> How long is the prompt? I don't think utf8 is the reason.
> 
> If tokens/sec is lower it may be because of the top-p sampling added recently (it is in PR enabled by default, but now disabled in the master branch, but not in utf8 branch). Try with -p 0 to disable it.

You where right it was not because of utf-8 I had my test-version where I had hardcoded the prompt to test, sorry! 🙈
But overall I'm a little concerned encoding about 170 tokens takes 2.8 seconds with runfast on my mac m2 🤔 not only #226

**Comment by atamurad** [comment]
@stellanhaglund 

https://github.com/karpathy/llama2.c/pull/251 improves encoding speed but it doesn't support UTF8.

I've implemented sorted vocabulary for faster lookup like in #251 but on top of utf8 branch. If you want to test, it's in my repository but I didn't update PR yet. https://github.com/atamurad/llama2.c/tree/utf8-sortedvocab

**Comment by stellanhaglund** [comment]
Really nice! 👍 works with my prompts and much faster!

**Comment by atamurad** [comment]
awesome, thanks for testing and confirmation of the fix. I'll update the PR to include this as well.

**Comment by karpathy** [comment]
This is now merged to master too now, closing.
